hedera-dedupe-bigquery/src/main/java/com/hedera/dedupe/DedupeRunner.java

```diff
@@ -0,0 +1,200 @@
+ * 2. Flag rows not in the streaming buffer: UPDATE table SET dedupe = 1 WHERE ...;
+ * 3. Get endTimestamp: SELECT MAX(consensusTimestamp) FROM table WHERE dedupe IS NOT NULL ....
+ * 4. Check for duplicates in [startTimestamp , endTimestamp] window
+ * 5. Remove duplicates (if present)
+ * 6. Save state: Set lastValidTimestamp in state table to endTimestamp
+ */
+@Log4j2
+@Component
+public class DedupeRunner {
+    private static final String JOB_NAME_FLAG_ROW = "flag_dml_accessible_rows";
+    private static final String JOB_NAME_GET_END_TIMESTAMP = "get_end_timestamp";
+    private static final String JOB_NAME_GET_DUPLICATES = "get_duplicates";
+    private static final String JOB_NAME_REMOVE_DUPLICATES = "remove_duplicates";
+
+    private final BigQueryHelper bigQueryHelper;
+    private final String transactionsTable;
+    private final DedupeState dedupeState;
+    private final DedupeMetrics metrics;
+
+    public DedupeRunner(DedupeProperties properties, BigQueryHelper bigQueryHelper, DedupeState dedupeState,
+                        DedupeMetrics dedupeMetrics) {
+        this.bigQueryHelper = bigQueryHelper;
+        this.dedupeState = dedupeState;
+        this.metrics = dedupeMetrics;
+        String datasetName = properties.getDatasetName();
+        bigQueryHelper.ensureTableExists(datasetName, properties.getTableName());
+        transactionsTable = properties.getProjectId() + "." + datasetName + "." + properties.getTableName();
+    }
+
+    @Scheduled(fixedRateString = "${hedera.dedupe.fixedRate:300000}") // default: 5 min
+    public void run() {
+        try {
+            Instant dedupeStart = Instant.now();
+            metrics.getInvocationsCounter().increment();
+            metrics.resetJobMetrics();
+            runDedupe();
+            metrics.getRuntimeGauge().set(Duration.between(dedupeStart, Instant.now()).getSeconds());
+        } catch (Exception e) {
+            log.error("Failed deduplication", e);
+            metrics.getFailuresCounter().increment();
+        }
+    }
+
+    private void runDedupe() throws Exception {
+        var state = dedupeState.getState();
+
+        // 1. Get state: startTimestamp = lastValidTimestamp + 1 (from stateTable). 0 if not set.
+        long startTimestamp = 0; // in nanos
+        if (state.containsKey(STATE_NAME_LAST_VALID_TIMESTAMP)) {
+            startTimestamp = state.get(STATE_NAME_LAST_VALID_TIMESTAMP).getLongValue() + 1;
+        }
+        metrics.getStartTimestampGauge().set(startTimestamp);
+
+        // 2. Flag rows not in the streaming buffer: UPDATE table SET dedupe = 1 WHERE ...;
+        setDedupeState(startTimestamp);
+
+        // 3. Get endTimestamp: SELECT MAX(consensusTimestamp) FROM table WHERE dedupe IS NOT NULL ....
+        long endTimestamp = getEndTimestamp(startTimestamp);
+        if (endTimestamp == startTimestamp) {
+            return;
+        }
+
+        // 4. Check for duplicates in [startTimestamp , endTimestamp] window
```

Comment:
A possible race condition between steps 4 and 5. It's so unlikely though, that we can probably ignore it.

Reply:
By race condition I mean a duplicate appearing after `hasDuplicates` is run. Another possibility is a duplicated appearing before `lastValidTimestamp`. Also very unlikely if we can guarantee our Dataflow job won't insert rows older than the oldest record in BigQuery's streaming buffer .

Reply:
Yes, that's the idea. It relies on the fact that it takes lot of time before streaming buffer gets written to DML accessible data (up to 90min), and that records won't be late by more than few seconds.
That all of course during normal functioning.
It doesn't guard against the following case:
If importer pushing txns to pubsub goes down for couple hours and publishes duplicate txns when it comes back up.
Or maybe if dataflow pipeline has some bug and gets stuck for hours.
Generally, given that we don't have operational expertise with this, i think it might be worthwhile to do a check every Xs for deduplicates in window [0 , now-X]  (and alert if any). X = 1 day would be good enough. What do you think?

Reply:
> it might be worthwhile to do a check every Xs for deduplicates in window [0 , now-X] (and alert if any). X = 1 day would be good enough. 

That's a good idea. We do this for Ethereum and Bitcoin datasets as well.

Reply:
Filed #13 .